staging/src/k8s.io/metrics/pkg/client/custom_metrics/client.go

```diff
@@ -20,27 +20,52 @@ import (
 	"fmt"
 
 	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	serializer "k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/util/flowcontrol"
+
+	cmint "k8s.io/metrics/pkg/apis/custom_metrics"
+	"k8s.io/metrics/pkg/apis/custom_metrics/v1beta1"
 	"k8s.io/metrics/pkg/apis/custom_metrics/v1beta2"
 	"k8s.io/metrics/pkg/client/custom_metrics/scheme"
 )
 
+var metricConverter = NewMetricConverter()
+var codecs = serializer.NewCodecFactory(metricConverter.Scheme())
+
+type AvailableMetricsAPIFunc func() (*metav1.APIGroupList, error)
+
+type APIVersionsFromDiscovery struct {
+	discovery.DiscoveryClient
+}
+
+func (d APIVersionsFromDiscovery) Versions() (*metav1.APIGroupList, error) {
+	groups, err := d.DiscoveryClient.ServerGroups()
+	if err != nil {
+		return nil, err
+	}
+
+	return groups, nil
+}
+
 type customMetricsClient struct {
-	client rest.Interface
-	mapper meta.RESTMapper
+	client          rest.Interface
+	mapper          meta.RESTMapper
+	metricsVersions AvailableMetricsAPIFunc
 }
 
-func New(client rest.Interface) CustomMetricsClient {
+func New(client rest.Interface, discoveryClient APIVersionsFromDiscovery) CustomMetricsClient {
```

Comment:
this should be taking a `AvailableMetricsAPIFunc`, not an `APIVersionsFromDiscovery`.

Reply:
ditto below